homeassistant/components/sensor/utility_meter.py

```diff
@@ -0,0 +1,270 @@
+"""
+Utility meter from sensors providing raw data.
+
+For more details about this platform, please refer to the documentation at
+https://home-assistant.io/components/sensor.utility_meter/
+"""
+import logging
+
+from decimal import Decimal
+import voluptuous as vol
+
+import homeassistant.util.dt as dt_util
+import homeassistant.helpers.config_validation as cv
+from homeassistant.components.sensor import (DOMAIN, PLATFORM_SCHEMA)
+from homeassistant.const import (
+    CONF_NAME, ATTR_UNIT_OF_MEASUREMENT, ATTR_ENTITY_ID,
+    EVENT_HOMEASSISTANT_START)
+from homeassistant.core import callback
+from homeassistant.helpers.event import (
+    async_track_state_change, async_track_time_change)
+from homeassistant.helpers.dispatcher import (
+    dispatcher_send, async_dispatcher_connect)
+from homeassistant.helpers.restore_state import RestoreEntity
+
+_LOGGER = logging.getLogger(__name__)
+
+ATTR_SOURCE_ID = 'source'
+ATTR_STATUS = 'status'
+ATTR_PERIOD = 'meter_period'
+ATTR_LAST_PERIOD = 'last_period'
+ATTR_LAST_RESET = 'last_reset'
+
+SIGNAL_START_PAUSE_METER = 'utility_meter_start_pause'
+SIGNAL_RESET_METER = 'utility_meter_reset'
+
+SERVICE_START_PAUSE = 'utility_meter_start_pause'
+SERVICE_RESET = 'utility_meter_reset'
+
+HOURLY = 'hourly'
+DAILY = 'daily'
+WEEKLY = 'weekly'
+MONTHLY = 'monthly'
+YEARLY = 'yearly'
+
+METER_TYPES = [HOURLY, DAILY, WEEKLY, MONTHLY, YEARLY]
+
+SERVICE_METER_SCHEMA = vol.Schema({
+    vol.Required(ATTR_ENTITY_ID): cv.entity_ids,
+})
+
+CONF_SOURCE_SENSOR = 'source'
+CONF_METER_TYPE = 'cycle'
+CONF_METER_OFFSET = 'offset'
+CONF_PAUSED = 'paused'
+
+ICON = 'mdi:counter'
+
+PRECISION = 3
+PAUSED = "paused"
+COLLECTING = "collecting"
+
+PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend({
+    vol.Required(CONF_SOURCE_SENSOR): cv.entity_id,
+    vol.Optional(CONF_NAME): cv.string,
+    vol.Optional(CONF_METER_TYPE): vol.In(METER_TYPES),
+    vol.Optional(CONF_METER_OFFSET, default=0): cv.positive_int,
+    vol.Optional(CONF_PAUSED, default=False): cv.boolean,
+})
+
+
+async def async_setup_platform(hass, config, async_add_entities,
+                               discovery_info=None):
+    """Set up the utility meter sensor."""
+    def run_setup(event):
+        """Delay the setup until Home Assistant is fully initialized."""
+        conf = discovery_info if discovery_info else config
+
+        meter = UtilityMeterSensor(hass,
+                                   conf[CONF_SOURCE_SENSOR],
+                                   conf.get(CONF_NAME),
+                                   conf.get(CONF_METER_TYPE),
+                                   conf.get(CONF_METER_OFFSET),
+                                   conf.get(CONF_PAUSED))
+
+        async_add_entities([meter])
+
+        @callback
+        def async_start_pause_meter(service):
+            """Process service start_pause meter."""
+            for entity in service.data.get(ATTR_ENTITY_ID):
```

Comment:
```suggestion
            for entity in service.data[ATTR_ENTITY_ID]:
```

`entity_id` is required.